include/openthread/tcp.h

```diff
@@ -0,0 +1,616 @@
+ * @retval OT_ERROR_FAILED  Failed to bind the TCP endpoint.
+ */
+otError otTcpBind(otInstance *aInstance, otTcpEndpoint *aEndpoint, const otSockAddr *aSockName);
+
+/**
+ * This enumeration defines flags passed to @p otTcpConnect.
+ */
+enum {
+    OT_TCP_CONNECT_NO_FAST_OPEN = 1 << 0,
+};
+
+/**
+ * Records the remote host and port for this connection. By default TCP Fast
+ * Open is used. This means that this function merely records the remote host
+ * and port, and that the TCP connection establishment handshake only happens
+ * on the first call to @p otTcpSendByReference. TCP Fast Open can be
+ * explicitly disabled using @p aFlags, in which case the TCP connection
+ * establishment handshake is initiated immediately.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the TCP endpoint belongs.
+ * @param[in]  aEndpoint  A pointer to the TCP endpoint structure to connect.
+ * @param[in]  aSockName  The IP address and port of the host to which to connect.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpConnect(otInstance *aInstance, otTcpEndpoint *aEndpoint, const otSockAddr *aSockName, uint32_t aFlags);
+
+/**
+ * This enumeration defines flags passed to @p otTcpSendByReference.
+ */
+enum {
+    OT_TCP_SEND_MORE_TO_COME = 1 << 0,
+};
+
+/**
+ * Adds data referenced by the linked buffer pointed to by @p aBuffer to the
+ * the send buffer. Upon a sucessful call to this function, the linked buffer
+ * and data it references are owned by the TCP stack; they should not be
+ * modified by the application until a "send done" callback returns ownership
+ * of those objects to the application. It is acceptable to call this function
+ * to add another linked buffer to the send queue, even if the "send done"
+ * callback for a previous invocation of this function has not yet fired.
+ *
+ * Note that @p aBuffer should not be chained; its @p mNext field should be
+ * @p NULL. If additional data will be added right after this call, then the
+ * @p OT_TCP_SEND_MORE_TO_COME flag should be used as a hint to the TCP
+ * implementation.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the TCP endpoint belongs.
+ * @param[in]  aEndpoint  A pointer to the TCP endpoint structure representing the TCP endpoint on which to send data.
+ * @param[in]  aBuffer    A pointer to the linked buffer chain referencing data to add to the send buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpSendByReference(otInstance *aInstance, otTcpEndpoint *aEndpoint, otLinkedBuffer *aBuffer, uint32_t aFlags);
```

Comment:
Let me add another idea/suggestion here to get people's thoughts on this.

I noticed we use `otInstance *` and `otTcpEndPoint *` as the paramters for the APIs and callbacks.

We can consider changing this as follows:
- We can assume a `otTcpEndPoint` would remember the `otInstance` it belongs to (it remembers it from a  successfully initialization from `otTcpEndpointInitialize()`.
- Then in all other APIs/callbacks we just use `otTcpEndPoint *` (and not pass the `otInstance *` anymore)
- We can also provide a new API for application to get the associated instance with an initialized `otTcpEndPoint`, something like:
```
otInstance *otTcpEndPointGetInstance(otTcpEndpoint *aEndpoint);
```
---------

I am myself somewhat on the fence on this:
- On one hand, we do use `otInstance *` as the first argument in almost all existing APIs (including UDP socket), so it may be good to keep it harmonized and same.
- On the other hand, I see the simplification that comes (ease of use) by just passing a `otTcpEndpoint *`.

Thoughts?

Reply:
Including the socket struct makes sense to me. Doing so will also reduce code size for each call point.

Reply:
@jwhui To clarify, you agree with both keeping the socket pointer and removing the context pointer from the argument list for callback functions. Right?

Reply:
> @jwhui To clarify, you agree with both keeping the socket pointer and removing the context pointer from the argument list for callback functions. Right?

Yes to both. Thanks!

Reply:
Good idea on extending to the callback context as well.
I also think it make sense to go for both.

Reply:
Done (see latest commit).